### PR TITLE
Adding support for Nebulas

### PR DESCRIPTION
**Pull request by Woo-ing**
## Description
Added Nebulas coin. 
https://nebulas.io/

## Testing instructions

This PR includes unit tests.

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.

### CODE REVIEW

File: TrustWalletCore.podspec [hunk 1]
```diff
@@ -18,7 +18,7 @@ Pod::Spec.new do |s|
     submodules: true
   }
 
-  protobuf_dir = 'build/protobuf/staging/protobuf-3.7.0'
+  protobuf_dir = 'build/protobuf/staging/protobuf-3.7.1'
```
Comment (hewigovens) [inline TrustWalletCore.podspec]: for upgrading protobuf, can you do that in a separate PR?

File: android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/nebulas/TestNebulasAddress.kt [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import wallet.core.jni.NebulasAddress
+
+class TestNebulasAddress {
+
+    init {
+        System.loadLibrary("TrustWalletCore")
+    }
+
+    @Test
+    fun testAddressFromPublicKey() {
```
Comment (hewigovens) [inline android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/nebulas/TestNebulasAddress.kt]: please also test address from plain string

File: android/trustwalletcore/build.gradle [hunk 3]
```diff
@@ -22,7 +22,7 @@ android {
     }
     externalNativeBuild {
         cmake {
-            version "3.10.2"
```
Comment (hewigovens) [inline android/trustwalletcore/build.gradle]: @eddywm could you please confirm? I remember we use a old cmake version here for some reasons

File: include/TrustWalletCore/TWNebulasAddress.h [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+#include "TWBase.h"
+#include "TWData.h"
+#include "TWString.h"
+
+TW_EXTERN_C_BEGIN
+
+struct TWPublicKey;
+
+/// Represents an Nebulas address.
```
Comment (hewigovens) [inline include/TrustWalletCore/TWNebulasAddress.h]:
```suggestion
/// Represents a Nebulas address.
```

File: include/TrustWalletCore/TWNebulasAddress.h [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+TW_EXPORT_STATIC_METHOD
+bool TWNebulasAddressIsValidString(TWString *_Nonnull string);
+
+/// Creates an address from a string representaion.
+TW_EXPORT_STATIC_METHOD
+struct TWNebulasAddress *_Nullable TWNebulasAddressCreateWithString(TWString *_Nonnull string);
+
+/// Creates an address from a key hash.
+TW_EXPORT_STATIC_METHOD
+struct TWNebulasAddress *_Nullable TWNebulasAddressCreateWithKeyHash(TWData *_Nonnull keyHash);
```
Comment (hewigovens) [inline include/TrustWalletCore/TWNebulasAddress.h]: where do we need to create a address from key hash? I think we can remove it

File: js/static-module [hunk 6]
```diff
@@ -0,0 +1,190 @@
+var TW.ARK.Proto = TW.ARK.Proto || exports || {}, exports;
```
Comment (hewigovens) [inline js/static-module]: please ignore this file

File: js/tests/blockchain/nebulas/NebulasAddress.test.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import { expect } from 'chai';
+import 'mocha';
+
+import { fromHexString } from '../../Utils';
+import { NebulasAddress, PublicKey, PublicKeyType, PrivateKey } from '../../../lib';
+
+describe('NebulasAddress', () => {
+
+    it('test address from PublicKey', () => {
```
Comment (hewigovens) [inline js/tests/blockchain/nebulas/NebulasAddress.test.ts]: same here, please test address from string

File: src/Nebulas/Address.cpp [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    if(data[0] != Address::AddressPrefix) {
+        return false;
+    }
+    if (data[1] != Address::NormalType && data[1] != Address::ContractType) {
+        return false;
+    }
+
+    Data content(data.begin(), data.begin()+22);
```
Comment (hewigovens) [inline src/Nebulas/Address.cpp]:
```suggestion
    Data content(data.begin(), data.begin() + 22);
```

Comment (hewigovens) [inline src/Nebulas/Address.cpp]:
```suggestion
    Data content(data.begin(), data.begin() + 22);
```

Comment (hewigovens) [reply src/Nebulas/Address.cpp]: please run `clang-format` for this file

File: src/Nebulas/Address.h [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+#include "../PublicKey.h"
+
+#include <array>
+#include <cstdint>
+#include <vector>
+
+namespace TW::Nebulas {
+
+class Address {
```
Comment (hewigovens) [inline src/Nebulas/Address.h]: we have a class called `Base58Address`, please reuse it if possible

File: src/Nebulas/Signer.cpp [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+const char* Signer::TxPayloadDeployType = "deploy";
+const char* Signer::TxPayloadCallType = "call";
+
+void Signer::sign(const PrivateKey& privateKey, Transaction& transaction) const noexcept {
+    auto hash = this->hash(transaction);
+    transaction.alg = 1;
+    transaction.sign = privateKey.sign(hash, TWCurveSECP256k1);
+}
+
+void Signer::appendBigEndian(Data& data, const uint256_t& value,uint32_t digit) {
```
Comment (hewigovens) [inline src/Nebulas/Signer.cpp]: can you tweak and move this function to `BinaryCoding.h`?

File: src/Nebulas/Transaction.cpp [hunk 11]
```diff
@@ -0,0 +1,9 @@
+// Copyright © 2017-2019 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#include "Transaction.h"
+
+using namespace TW;
```
Comment (hewigovens) [inline src/Nebulas/Transaction.cpp]: maybe we can remove this file?

File: src/proto/Nebulas.proto [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    bytes gas_price = 4;
+
+    
+    // Gas limit (256-bit number)
+    bytes gas_limit = 5;
+
+    // Recipient's address.
+    string to_address = 6;
+
+    // Amount to send in wei (256-bit number)
```
Comment (hewigovens) [inline src/proto/Nebulas.proto]: `wei`(ethereum?) or other unit?

File: src/proto/Nebulas.proto [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    // Optional payload
+    bytes payload = 9;
+
+    // Private key.
+    bytes private_key = 10;
+}
+
+// Transaction signing output.
+message SigningOutput {
+    uint32 alg = 1;
```
Comment (hewigovens) [inline src/proto/Nebulas.proto]: what's alg short for? prefer a longer name

File: src/proto/Nebulas.proto [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+    bytes sign = 2;
```
Comment (hewigovens) [inline src/proto/Nebulas.proto]: sign -> signature?

File: tests/Nebulas/AddressTests.cpp [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+TEST(NebulasAddress, FromPrivateKey) {
+    const auto privateKey = PrivateKey(parse_hex("d2fd0ec9f6268fc8d1f563e3e976436936708bdf0dc60c66f35890f5967a8d2b"));
+    const auto publicKey = PublicKey(privateKey.getPublicKey(TWPublicKeyTypeSECP256k1Extended));
+    const auto address = Address(publicKey);
+
+    ASSERT_EQ(address.string(), "n1V5bB2tbaM3FUiL4eRwpBLgEredS5C2wLY");
+}
+
+TEST(NebulasAddress, IsValid) {
+    ASSERT_FALSE(Address::isValid("abc"));
```
Comment (hewigovens) [inline tests/Nebulas/AddressTests.cpp]: this line is tested in `Invalid` case, address starts with `n1` is normal address, can you also test `ContractType` address?

File: TrustWalletCore.podspec (hunk 1 above)
Comment (Woo-ing) [reply TrustWalletCore.podspec]: ok, I will revert it  to 3.7.0.

File: android/trustwalletcore/build.gradle (hunk 3 above)
Comment (Woo-ing) [reply android/trustwalletcore/build.gradle]: ok, I'd revert it back.

File: include/TrustWalletCore/TWNebulasAddress.h (hunk 5 above)
Comment (Woo-ing) [reply include/TrustWalletCore/TWNebulasAddress.h]: yes,  I removed it.

File: js/static-module (hunk 6 above)
Comment (Woo-ing) [reply js/static-module]: ok, I will rm this file from git.

File: js/tests/blockchain/nebulas/NebulasAddress.test.ts (hunk 7 above)
Comment (Woo-ing) [reply js/tests/blockchain/nebulas/NebulasAddress.test.ts]: ok

File: src/Nebulas/Signer.cpp (hunk 10 above)
Comment (Woo-ing) [reply src/Nebulas/Signer.cpp]: Ok, but I think it should move to uint256.h. It is more like the **store()** function.

File: src/Nebulas/Transaction.cpp (hunk 11 above)
Comment (Woo-ing) [reply src/Nebulas/Transaction.cpp]: It's no problem.

File: src/proto/Nebulas.proto (hunk 12 above)
Comment (Woo-ing) [reply src/proto/Nebulas.proto]: 1 NAS = 10^18 Wei, It's just like Ethereum; :)

File: src/proto/Nebulas.proto (hunk 13 above)
Comment (Woo-ing) [reply src/proto/Nebulas.proto]: It's means algorithm type. only one value now(1=SECP256K1), perhaps it has more value in the future.

File: src/proto/Nebulas.proto (hunk 14 above)
Comment (Woo-ing) [reply src/proto/Nebulas.proto]: Yes, I will renamed it.

File: tests/Nebulas/AddressTests.cpp (hunk 15 above)
Comment (Woo-ing) [reply tests/Nebulas/AddressTests.cpp]: Ok, that's a good idea!

File: src/Nebulas/Address.cpp (hunk 8 above)
Comment (Woo-ing) [reply src/Nebulas/Address.cpp]: ok

Comment (Woo-ing) [reply src/Nebulas/Address.cpp]: I have used the Clang-format for this file.

File: js/tests/blockchain/nebulas/NebulasAddress.test.ts (hunk 7 above)
Comment (Woo-ing) [reply js/tests/blockchain/nebulas/NebulasAddress.test.ts]: ok

File: js/static-module (hunk 6 above)
Comment (Woo-ing) [reply js/static-module]: ok

File: include/TrustWalletCore/TWNebulasAddress.h (hunk 5 above)
Comment (Woo-ing) [reply include/TrustWalletCore/TWNebulasAddress.h]: I have removed this function.

File: include/TrustWalletCore/TWNebulasAddress.h (hunk 4 above)
Comment (Woo-ing) [reply include/TrustWalletCore/TWNebulasAddress.h]: thanks.

File: android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/nebulas/TestNebulasAddress.kt (hunk 2 above)
Comment (Woo-ing) [reply android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/nebulas/TestNebulasAddress.kt]: ok.

File: src/Nebulas/Address.h (hunk 9 above)
Comment (Woo-ing) [reply src/Nebulas/Address.h]: Ok, Nebulas::Address is inherited from Base58Address now.

File: src/Nebulas/Signer.cpp [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+using namespace TW;
+using namespace TW::Nebulas;
+
+const char *Signer::TxPayloadBinaryType = "binary";
+const char *Signer::TxPayloadDeployType = "deploy";
+const char *Signer::TxPayloadCallType = "call";
+
+void Signer::sign(const PrivateKey &privateKey, Transaction &transaction) const noexcept {
+    auto hash = this->hash(transaction);
+    transaction.algorithm = 1;
```
Comment (hewigovens) [inline src/Nebulas/Signer.cpp]: can you use value from `SigningInput`? not hardcoded

File: src/Nebulas/Transaction.h [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+    uint256_t gasPrice;
+    uint256_t gasLimit;
+    Address to;
+    uint256_t amount;
+    uint256_t timestamp;
+    std::vector<uint8_t> payload;
+
+    // Signature values
+    Data signature;
+    uint32_t algorithm=1;
```
Comment (hewigovens) [inline src/Nebulas/Transaction.h]: use value from input

File: src/uint256.h [hunk 18]
```diff
@@ -48,4 +48,19 @@ inline Data store(const uint256_t& v) {
     return bytes;
 }
 
+// Append a uint256_t value as a big-endian byte array into the provided buffer, and limit
+// the array size by digit/8.
+inline void appendBigEndian(Data &data, const uint256_t &value, uint32_t digit) {
```
Comment (hewigovens) [inline src/uint256.h]: for consistency, how about `appendBigEndian` -> `encode256BE(value, digit, data)`?

File: swift/protobuf [hunk 19]
```diff
@@ -1 +1 @@
-../build/protobuf/staging/protobuf-3.7.0/src
\ No newline at end of file
+../build/protobuf/staging/protobuf-3.7.0/src/
```
Comment (hewigovens) [inline swift/protobuf]: can you revert this file?

File: android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/nebulas/TestNebulasSigner.kt [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+            gasLimit = ByteString.copyFrom("0x030d40".toHexByteArray()) //200000
+            amount = ByteString.copyFrom("0x98a7d9b8314c0000".toHexByteArray())    //11000000000000000000
+            timestamp = ByteString.copyFrom("0x5cfc84ca".toHexByteArray())   //1560052938
+            payload = ByteString.copyFrom(byteArrayOf(10,6,98,105,110,97,114,121))
+            privateKey = ByteString.copyFrom(PrivateKey("d2fd0ec9f6268fc8d1f563e3e976436936708bdf0dc60c66f35890f5967a8d2b".toHexByteArray()).data())
+        }
+
+        val output: Nebulas.SigningOutput = NebulasSigner.sign(signingInput.build())
+
+        assertEquals(output.algorithm,1)
```
Comment (hewigovens) [inline android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/nebulas/TestNebulasSigner.kt]:
```suggestion
        assertEquals(output.algorithm, 1)
```

File: swift/Tests/Blockchains/NebulasTests.swift [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+        let input = NebulasSigningInput.with {
+            $0.chainID = Data(hexString: "01")!
+            $0.fromAddress = "n1V5bB2tbaM3FUiL4eRwpBLgEredS5C2wLY"
+            $0.nonce = Data(hexString: "07")!
+            $0.gasPrice = Data(hexString: "0f4240")!     //1000000
+            $0.gasLimit = Data(hexString: "030d40")!     //200000
+            $0.toAddress = "n1SAeQRVn33bamxN4ehWUT7JGdxipwn8b17"
+            $0.amount = Data(hexString: "98a7d9b8314c0000")!  //11000000000000000000ULL
+            $0.timestamp = Data(hexString: "5cfc84ca")!   //1560052938
+            $0.payload = Data(bytes: arr, count: 8)
```
Comment (hewigovens) [inline swift/Tests/Blockchains/NebulasTests.swift]: I think you can just use `Data(hexString:)`?

File: src/Nebulas/Signer.cpp (hunk 16 above)
Comment (Woo-ing) [reply src/Nebulas/Signer.cpp]: Ok, I will add a new function use the value by SigningInput.

File: src/Nebulas/Transaction.h (hunk 17 above)
Comment (Woo-ing) [reply src/Nebulas/Transaction.h]: Ok, I removed this default value.

File: src/uint256.h (hunk 18 above)
Comment (Woo-ing) [reply src/uint256.h]: It's ok.

File: swift/protobuf (hunk 19 above)
Comment (Woo-ing) [reply swift/protobuf]: Ok, I have reverted it.

File: android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/nebulas/TestNebulasSigner.kt (hunk 20 above)
Comment (Woo-ing) [reply android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/nebulas/TestNebulasSigner.kt]: ok

File: swift/Tests/Blockchains/NebulasTests.swift (hunk 21 above)
Comment (Woo-ing) [reply swift/Tests/Blockchains/NebulasTests.swift]: Let me try.

Comment (Woo-ing) [reply swift/Tests/Blockchains/NebulasTests.swift]: Can I convert the payload from a string? that's more understandably.

File: src/Tron/Protobuf/TronInternal.pb.h [hunk 22]
```diff
@@ -13,7 +13,7 @@
 #error incompatible with your Protocol Buffer headers. Please update
 #error your headers.
 #endif
-#if 3007000 < PROTOBUF_MIN_PROTOC_VERSION
```
Comment (hewigovens) [inline src/Tron/Protobuf/TronInternal.pb.h]: could you please also revert Tron/Zilliqa pb.h?

Comment (Woo-ing) [reply src/Tron/Protobuf/TronInternal.pb.h]: oh, I have merged the official master branch. Should I roll back?

Comment (Woo-ing) [reply src/Tron/Protobuf/TronInternal.pb.h]: I have revert these file.
